[saw-remote-api] [test] `tests/saw/test_salsa20.py`: `ExpandContract` should allocate pointer arguments as read-only

If we change the `size` to `65` (or more) when verifying that `s20_crypt32` meets `Salsa20CryptContract`...

```python
class Salsa20EasyTest(unittest.TestCase):
    def test_salsa20(self):
        # ...
        crypt_result = llvm_verify(mod, 's20_crypt32', Salsa20CryptContract(65), lemmas=[expand_result])
        # ...
```

..., verification fails due to a failed override assumption:

```
saw-script/saw-remote-api/python$ poetry run python tests/saw/test_salsa20.py
...

⚠️  Failed to verify: lemma_Salsa20CryptContract (defined at .../saw-script/saw-remote-api/python/tests/saw/test_salsa20.py:110):
error: user error (Symbolic execution failed.
Abort due to false assumption:
  salsa20.c:182:7: error: in s20_crypt32
  All overrides failed during structural matching:
  *  Name: s20_expand32
     Location: internal
     Argument types:
     - i8*
     - i8*
     - i8*
     Return type: <void>
     Arguments:
     - concrete pointer: allocation = 15, offset = 0
     - concrete pointer: allocation = 25, offset = 0
     - concrete pointer: allocation = 24, offset = 0
     at internal
     error when loading through pointer that appeared in the override's points-to precondition(s):
     Precondition:
       Pointer: concrete pointer: allocation = 25, offset = 0
       Pointee: fresh:x0#670 : [16][8]

       Assertion made at: internal
     Failure reason:
       When reading through pointer: (25, 0x0:[64])
       in the  precondition of an override
       Tried to read something of size: 16
       And type: [16 x i8]
       Found 1 possibly matching allocation(s):
       - StackAlloc 25 0x10:[64] Mutable 16-byte-aligned salsa20.c:155:11
Stack frame s20_crypt32
```

Amending `ExpandContract` to allocate memory for pointer arguments as read-only...

```python
class ExpandContract(Contract):
    def specification(self):
        # ...
        k_p = self.alloc(LLVMArrayType(i8, 32), read_only = True)  # <--
        n_p = self.alloc(LLVMArrayType(i8, 16), read_only = True)  # <--
        # ...
```

...corrects this:

```
...

✅  Verified: lemma_Salsa20CryptContract (defined at .../saw-script/saw-remote-api/python/tests/saw/test_salsa20.py:110)

...
```

--- a/saw-remote-api/python/tests/saw/test_salsa20.py
+++ b/saw-remote-api/python/tests/saw/test_salsa20.py
@@ -7,7 +7,7 @@ from saw_client.llvm import Contract, void, SetupVal, FreshVar, cryptol, i8, i32
 
 def ptr_to_fresh(c : Contract, ty : LLVMType, name : Optional[str] = None) -> Tuple[FreshVar, SetupVal]:
     """Add to``Contract`` ``c`` an allocation of a pointer of type ``ty`` initialized to an unknown fresh value.
-    
+
     :returns A fresh variable bound to the pointers initial value and the newly allocated pointer. (The fresh
              variable will be assigned ``name`` if provided/available.)"""
     var = c.fresh_var(ty, name)
@@ -15,7 +15,7 @@ def ptr_to_fresh(c : Contract, ty : LLVMType, name : Optional[str] = None) -> Tu
     return (var, ptr)
 
 def oneptr_update_func(c : Contract, ty : LLVMType, fn_name : str) -> None:
-    """Upcates contract ``c`` to declare calling it with a pointer of type ``ty``
+    """Updates contract ``c`` to declare calling it with a pointer of type ``ty``
     updates that pointer with the result, which is equal to calling the
     Cryptol function ``fn_name``."""
     (x, x_p) = ptr_to_fresh(c, ty)
@@ -92,8 +92,8 @@ class ExpandContract(Contract):
     def specification(self):
         k = self.fresh_var(LLVMArrayType(i8, 32))
         n = self.fresh_var(LLVMArrayType(i8, 16))
-        k_p = self.alloc(LLVMArrayType(i8, 32))
-        n_p = self.alloc(LLVMArrayType(i8, 16))
+        k_p = self.alloc(LLVMArrayType(i8, 32), read_only = True)
+        n_p = self.alloc(LLVMArrayType(i8, 16), read_only = True)
         ks_p = self.alloc(LLVMArrayType(i8, 64))
         self.points_to(k_p, k)
         self.points_to(n_p, n)
@@ -117,8 +117,8 @@ class Salsa20CryptContract(Contract):
 
         self.execute_func(k_p, v_p, cryptol('0 : [32]'), m_p, cryptol(f'{self.size!r} : [32]'))
 
-        self.returns(cryptol('0 : [32]'))
         self.points_to(m_p, cryptol("Salsa20_encrypt")((k, v, m)))
+        self.returns(cryptol('0 : [32]'))
 
 class Salsa20EasyTest(unittest.TestCase):
     def test_salsa20(self):
@@ -146,8 +146,12 @@ class Salsa20EasyTest(unittest.TestCase):
         self.assertIs(hash_result.is_success(), True)
         expand_result = llvm_verify(mod, 's20_expand32', ExpandContract(), lemmas=[hash_result])
         self.assertIs(expand_result.is_success(), True)
-        crypt_result = llvm_verify(mod, 's20_crypt32', Salsa20CryptContract(63), lemmas=[expand_result])
-        self.assertIs(crypt_result.is_success(), True)
+        crypt_result_63 = llvm_verify(mod, 's20_crypt32', Salsa20CryptContract(63), lemmas=[expand_result])
+        self.assertIs(crypt_result_63.is_success(), True)
+        crypt_result_64 = llvm_verify(mod, 's20_crypt32', Salsa20CryptContract(64), lemmas=[expand_result])
+        self.assertIs(crypt_result_64.is_success(), True)
+        crypt_result_65 = llvm_verify(mod, 's20_crypt32', Salsa20CryptContract(65), lemmas=[expand_result])
+        self.assertIs(crypt_result_65.is_success(), True)
 
 if __name__ == "__main__":
     unittest.main()

